### DISCUSSION TITLE
How to vi.mock only in one test

### DISCUSSION THREAD

**Discussion by Hurtak** [vitest-dev/vitest#2511] [Q&A]
I am using the `vi.mock` to mock module but only in one test.

I tried the following but it does not seem to work, the `original` test still has the mocked module data

```tsx
import { vi, test, expect, describe, afterEach } from 'vitest';
import { data } from './fetch';

describe('fetch', () => {
  afterEach(() => {
    vi.clearAllMocks();
    vi.resetAllMocks();
  });

  test('mock', async () => {
    vi.mock('node-fetch', () => ({
      default: () => Promise.resolve({ json: () => Promise.resolve({ mock: 'data' }) }),
    }));
    expect(await data()).toEqual({ mock: 'data' });
  });

  test('original', async () => {
    expect(await data()).toEqual({
      id: 1,
      todo: 'Do something nice for someone I care about',
      completed: true,
      userId: 26,
    });
  });
});

```

Any solution to this? Or am I using the vi.mock incorrectly?

**Comment by sheremet-va** [comment]
>  Or am I using the vi.mock incorrectly?

[`vi.mock` is hoisted to top of the file.](https://vitest.dev/api/#vi-mock):

> The call to vi.mock is hoisted, so it doesn't matter where you call it. It will always be executed before all imports.

You can check mocking cheat sheet for different ways to mock a module: https://vitest.dev/guide/mocking.html#cheat-sheet

**Reply by Hurtak** [reply]
I was checking that but I haven't really found a simple solution that would work well.

**Comment by sandra-oldenburg** [comment]
I achieved it like this for mocking a component. 

 ```
import { vi, it, describe } from 'vitest';
import * as childComponent from './path/to/child.tsx'

 describe('Parent', () => {
      it('tests with mocked child', () => {
           vi.spyOn(childComponent, 'Child').mockImplementation((props: ChildProps) => {
                return (
                     <div>
                          // your component mock able to use component props as well
                     </div>
                )
      })
 })
```

**Comment by AriPerkkio** [comment]
Here's another approach:
https://stackblitz.com/~/edit/vitest-dev-vitest-afppg3?file=test/basic.test.ts

**Reply by longzheng** [reply]
Ah great idea.

I modified it slightly to allow each test to mock the implementation perpetually (instead of just once), which gets reset with a `beforeEach`

```diff
vi.mock(import('../src/math.js'), async (importOriginal) => {
  const og = await importOriginal();
- return { ...og, sum: vi.fn().mockImplementation(og.sum) }
+ return { ...og, sum: vi.fn(og.sum) }
})

+ beforeEach(() => {
+   vi.restoreAllMocks();
+ });

test('works when mocked', () => {
-  vi.mocked(math.sum).mockImplementationOnce((a, b) => a + 100)
+  vi.mocked(math.sum).mockImplementation((a, b) => a + 100)
  
  expect(wrapper(4, 4)).toBe(104)
})
```

Example https://stackblitz.com/~/edit/vitest-dev-vitest-gmvw67?file=test/basic.test.ts

**Comment by JesusP2** [comment]
I guess you can use vi.doMock now: https://vitest.dev/api/vi.html#vi-domock
